src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeBuilder.cs

```diff
@@ -97,45 +97,90 @@ public void AddContent(int sequence, RenderFragment fragment)
+        /// The attribute is associated with the most recently added element. If the value is <c>null</c> and the
+        /// current element is not a component, the frame will be omitted.
         /// </summary>
         /// <param name="sequence">An integer that represents the position of the instruction in the source code.</param>
         /// <param name="name">The name of the attribute.</param>
         /// <param name="value">The value of the attribute.</param>
         public void AddAttribute(int sequence, string name, UIEventHandler value)
         {
             AssertCanAddAttribute();
-            Append(RenderTreeFrame.Attribute(sequence, name, value));
+            if (value != null || _lastNonAttributeFrameType == RenderTreeFrameType.Component)
+            {
+                Append(RenderTreeFrame.Attribute(sequence, name, value));
+            }
         }
 
         /// <summary>
         /// Appends a frame representing a string-valued attribute.
-        /// The attribute is associated with the most recently added element.
+        /// The attribute is associated with the most recently added element. If the value is <c>null</c>, or
+        /// the <see cref="System.Boolean" /> value <c>false</c> and the current element is not a component, the 
+        /// frame will be omitted.
         /// </summary>
         /// <param name="sequence">An integer that represents the position of the instruction in the source code.</param>
         /// <param name="name">The name of the attribute.</param>
         /// <param name="value">The value of the attribute.</param>
         public void AddAttribute(int sequence, string name, object value)
         {
+            // This looks a bit daunting because we need to handle the boxed/object version of all of the
+            // types that AddAttribute special cases.
             if (_lastNonAttributeFrameType == RenderTreeFrameType.Element)
             {
-                // Element attribute values can only be strings or UIEventHandler
-                Append(RenderTreeFrame.Attribute(sequence, name, value.ToString()));
+                if (value is null)
+                {
+                    // Do nothing, treat 'null' attribute values for elements as a conditional attribute.
+                }
+                else if (value is bool boolValue)
+                {
+                    // bools have special behavior for an element.
+                    Append(RenderTreeFrame.Attribute(sequence, name, boolValue));
+                }
+                else if (value is UIEventHandler eventHandler)
+                {
+                    Append(RenderTreeFrame.Attribute(sequence, name, eventHandler));
+                }
+                else
+                {
+                    // Element attribute values can only be strings or UIEventHandler
+                    Append(RenderTreeFrame.Attribute(sequence, name, value.ToString()));
+                }
```

Comment:
These cases weren't handled before. If you had a weakly typed event handler, it would all go wrong for you.

Reply:
Ah, great - well spotted!